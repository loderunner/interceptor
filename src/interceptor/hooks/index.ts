export { useIntercepts, useIntercept } from './intercept'
export { RequestProvider, useRequests, useRequest } from './request'
