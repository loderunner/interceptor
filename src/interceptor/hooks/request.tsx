import {
  Provider as JotaiProvider,
  atom,
  useAtomValue,
  useSetAtom,
} from 'jotai'
import React, { useLayoutEffect } from 'react'

import * as Interceptor from '..'

const requestScope = Symbol('RequestProviderScope')

interface Props {
  children: React.ReactNode
}

const requestsAtom = atom<Interceptor.Request[] | undefined>(undefined)

// Children wrapper to use jotai hooks inside Provider scope
const Children = ({ children }: Props) => {
  const setRequests = useSetAtom(requestsAtom, requestScope)
  useLayoutEffect(() => {
    return Interceptor.subscribe(() => {
      setRequests([...Interceptor.requests])
    })
  }, [])
  return <>{children}</>
}

export const RequestProvider = ({ children }: Props) => {
  return (
    <JotaiProvider
      scope={requestScope}
      initialValues={[[requestsAtom, [...Interceptor.requests]]]}
    >
      <Children>{children}</Children>
    </JotaiProvider>
  )
}

export const useRequests = () => {
  const requests = useAtomValue(requestsAtom, requestScope)

  if (requests === undefined) {
    throw new Error(
      'no requests in scope\nDid you call useRequests inside a RequestProvider?'
    )
  }

  return requests as ReadonlyArray<Readonly<Interceptor.Request>>
}

export const useRequest = (id: string) => {
  const requests = useAtomValue(requestsAtom, requestScope)
  if (requests === undefined) {
    throw new Error(
      'no requests in scope\nDid you call useRequest inside a RequestProvider?'
    )
  }

  const request = requests.find((r) => r.id === id)
  if (request === undefined) {
    throw new Error('request not found')
  }

  return request as Readonly<Interceptor.Request>
}
