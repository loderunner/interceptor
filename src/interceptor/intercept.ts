export interface Intercept {
  pattern?: string
  enabled: boolean
}

export const intercepts: Intercept[] = []

export const addIntercept = (intercept: Intercept) => {
  intercepts.push(intercept)
}

export const removeIntercept = (intercept: Intercept) => {
  const i = intercepts.indexOf(intercept)
  if (i !== -1) {
    intercepts.splice(i, 1)
  }
}
