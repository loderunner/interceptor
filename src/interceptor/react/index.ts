export { Provider as InterceptProvider, useIntercepts } from './intercept'
