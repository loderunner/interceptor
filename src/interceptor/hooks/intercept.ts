import { atom, useAtom, useAtomValue } from 'jotai'
import { splitAtom } from 'jotai/utils'
import { useCallback } from 'react'

import * as Interceptor from '../intercept'

import type { Intercept } from '../intercept'

const interceptsAtom = atom([...Interceptor.intercepts])

export const useIntercepts = () => {
  const [intercepts, setIntercepts] = useAtom(interceptsAtom)

  const addIntercept = useCallback((inter: Intercept): Intercept => {
    const ret = Interceptor.addIntercept(inter)
    setIntercepts([...Interceptor.intercepts])
    return ret
  }, [])

  const removeIntercept = useCallback((id: string) => {
    Interceptor.removeIntercept(id)
    setIntercepts([...Interceptor.intercepts])
  }, [])

  return {
    intercepts: intercepts as ReadonlyArray<Readonly<Intercept>>,
    addIntercept,
    removeIntercept,
  }
}

const interceptAtomsAtom = splitAtom(interceptsAtom)

export const useIntercept = (id: string) => {
  for (const [i, inter] of Interceptor.intercepts.entries()) {
    if (inter.id === id) {
      const interceptAtoms = useAtomValue(interceptAtomsAtom)
      const interceptAtom = interceptAtoms[i]
      const [intercept, setIntercept] = useAtom(interceptAtom)
      return {
        intercept: intercept as Readonly<Intercept>,
        setIntercept: (newInter: Omit<Intercept, 'id'>) => {
          const inter = Interceptor.updateIntercept(id, newInter)
          if (inter === undefined) {
            throw new Error(`intercept '${id}' not found`)
          }
          setIntercept({ ...inter })
        },
      }
    }
  }

  throw new Error('intercept not found')
}
