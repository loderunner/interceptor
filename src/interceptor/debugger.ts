import { intercepts } from './intercept'
import { pushRequest } from './request'

import type { Protocol } from 'devtools-protocol'

type TargetInfo = chrome.debugger.TargetInfo
type Debuggee = chrome.debugger.Debuggee
export type RequestPausedEvent = Protocol.Fetch.RequestPausedEvent

let pausedFlag = false
export const paused = () => pausedFlag
export const pause = () => (pausedFlag = true)
export const unpause = () => (pausedFlag = false)

const onDebuggerEvent = (
  target: Debuggee,
  method: string,
  params?: unknown
) => {
  if (method === 'Fetch.requestPaused') {
    const event: RequestPausedEvent = params as RequestPausedEvent
    if (!pausedFlag) {
      for (const inter of intercepts) {
        // Ignore disabled intercept
        if (!inter.enabled) {
          continue
        }

        // Ignore empty intercepts
        if (inter.pattern === undefined || inter.pattern === '') {
          continue
        }

        // Dispatch event if the intercept matches
        // then break from loop to avoid duplicate capture
        const req = event.request
        let match = false
        if (inter.regexp) {
          let re: RegExp
          try {
            re = new RegExp(inter.pattern, 'i')
          } catch (error) {
            continue
          }
          match = re.test(req.url)
        } else {
          match = req.url.includes(inter.pattern)
        }
        if (match) {
          pushRequest({ id: event.requestId, ...req })
          break
        }
      }
    }

    chrome.debugger.sendCommand(target, 'Fetch.continueRequest', {
      requestId: event.requestId,
    })
  }
}

let debuggee: chrome.debugger.Debuggee | undefined

const getTargets = async () => {
  return new Promise<TargetInfo[]>((resolve) =>
    chrome.debugger.getTargets((targets) => resolve(targets))
  )
}

export const listen = async () => {
  if (debuggee !== undefined) {
    return
  }

  const targets = await getTargets()
  const target = targets.find(
    (t) => t.tabId === chrome.devtools.inspectedWindow.tabId
  )

  if (target === undefined) {
    throw new Error("Could not subscribe to inspected window's requests")
  }

  debuggee = { targetId: target.id }
  await chrome.debugger.attach(debuggee, '1.3')
  chrome.debugger.onEvent.addListener(onDebuggerEvent)
  await chrome.debugger.sendCommand(debuggee, 'Fetch.enable', {})
}

export const unlisten = () => {
  if (debuggee === undefined) {
    return
  }

  chrome.debugger.sendCommand(debuggee, 'Fetch.disable', {})
  chrome.debugger.onEvent.removeListener(onDebuggerEvent)
  chrome.debugger.detach(debuggee)
  debuggee = undefined
}
