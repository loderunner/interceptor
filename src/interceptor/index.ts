import { Intercept, intercepts as readWriteIntercepts } from './intercept'
import { Request, requests as readWriteRequests } from './request'

const readonlyIntercepts = readWriteIntercepts as ReadonlyArray<
  Readonly<Intercept>
>
export const intercepts = readonlyIntercepts

const readonlyRequests = readWriteRequests as ReadonlyArray<Readonly<Request>>
export const requests = readonlyRequests

export { listen, unlisten } from './debugger'
export {
  Request,
  RequestEventListener,
  subscribe,
  continueRequest,
} from './request'
export { Intercept, addIntercept, removeIntercept } from './intercept'
