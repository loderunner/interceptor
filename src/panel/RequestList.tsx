import * as React from 'react'
import { useMemo } from 'react'

import * as Interceptor from '@/interceptor'

import List from './components/List'

interface ItemProps {
  request: Interceptor.Request
}

const Item = ({ request }: ItemProps) => (
  <div className="select-none overflow-hidden text-ellipsis whitespace-nowrap">
    <span>{request.url}</span>
  </div>
)

interface Props {
  className?: string
  requests: Interceptor.Request[]
}

const RequestList = ({ className, requests }: Props) => {
  const header = useMemo(
    () => (
      <div className="flex select-none justify-between bg-slate-100 p-1 font-bold">
        Requests
      </div>
    ),
    []
  )

  const items = useMemo(
    () => requests.map((req, i) => <Item key={i} request={req} />),
    [requests]
  )

  return (
    <List
      id="request-list"
      className={className}
      header={header}
      items={items}
    />
  )
}

export default RequestList
