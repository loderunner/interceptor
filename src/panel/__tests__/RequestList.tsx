import { cleanup, render } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import { RequestProvider } from '@/interceptor/hooks'

import RequestList from '../RequestList'

describe('[RequestList]', () => {
  afterEach(() => {
    cleanup()
    jest.resetAllMocks()
  })

  it('should match empty snapshot', () => {
    const { container } = render(
      <RequestProvider>
        <RequestList />
      </RequestProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should match requestful snapshot', async () => {
    const { container } = render(
      <RequestProvider>
        <RequestList />
      </RequestProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
