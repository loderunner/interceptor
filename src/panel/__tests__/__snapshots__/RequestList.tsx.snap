// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[RequestList] should match requestful snapshot 1`] = `
<div>
  <div
    class="flex flex-col "
    id="request-list"
    role="list"
  >
    <div
      class="p-1 flex space-x-1 justify-between select-none bg-slate-100"
    >
      <span
        class="font-bold flex-auto"
      >
        Requests
      </span>
      <button
        class="self-stretch"
        title="Continue all requests"
      >
        <svg
          class="h-full w-auto rotate-90"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-visible"
    >
      <div
        class="px-1 flex select-none overflow-hidden "
        role="listitem"
      >
        <span
          class="flex-auto mr-1 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          https://www.example.com/complete/search?q=toto&client=gws-wiz
        </span>
        <button
          class="self-stretch"
          title="Cancel request"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
        <button
          class="self-stretch"
          title="Continue request"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`[RequestList] should match responseful snapshot 1`] = `
<div>
  <div
    class="flex flex-col "
    id="request-list"
    role="list"
  >
    <div
      class="p-1 flex space-x-1 justify-between select-none bg-slate-100"
    >
      <span
        class="font-bold flex-auto"
      >
        Requests
      </span>
      <button
        class="self-stretch"
        title="Continue all requests"
      >
        <svg
          class="h-full w-auto rotate-90"
        />
      </button>
    </div>
    <div
      class="flex flex-col overflow-y-visible"
    >
      <div
        class="px-1 flex select-none overflow-hidden "
        role="listitem"
      >
        <span
          class="self-stretch"
        >
          <svg
            class="h-full w-auto fill-green-600"
          />
        </span>
        <span
          class="flex-auto mr-1 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          https://www.example.com/complete/search?q=toto&client=gws-wiz
        </span>
        <button
          class="self-stretch"
          title="Cancel request"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
        <button
          class="self-stretch"
          title="Continue request"
        >
          <svg
            class="h-full w-auto"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
